File: lib-esphome-client/src/main/java/io/github/arhor/esphome/client/async/noise/NoiseCipherState.java
package io.github.arhor.esphome.client.async.noise;

import io.github.arhor.esphome.client.async.exception.EspHomeProtocolException;

import javax.crypto.Cipher;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.security.GeneralSecurityException;

public final class NoiseCipherState {

    private byte[] key;
    private long nonce;

    public void initializeKey(final byte[] newKey) {
        this.key = newKey == null ? null : newKey.clone();
        nonce = 0;
    }

    public boolean hasKey() {
        return key != null;
    }

    public byte[] encryptWithAd(final byte[] associatedData, final byte[] plaintext) {
        if (key == null) {
            return plaintext;
        }
        if (plaintext.length + NoiseConstants.AUTH_TAG_LENGTH > NoiseConstants.NOISE_MAX_MESSAGE_LENGTH) {
            throw new EspHomeProtocolException("Noise plaintext is too large: " + plaintext.length + " bytes");
        }
        return runCipher(Cipher.ENCRYPT_MODE, associatedData, plaintext);
    }

    public byte[] decryptWithAd(final byte[] associatedData, final byte[] ciphertext) {
        if (key == null) {
            return ciphertext;
        }
        if (ciphertext.length < NoiseConstants.AUTH_TAG_LENGTH) {
            throw new EspHomeProtocolException("Noise ciphertext is too short");
        }
        return runCipher(Cipher.DECRYPT_MODE, associatedData, ciphertext);
    }

    private byte[] runCipher(final int mode, final byte[] associatedData, final byte[] input) {
        try {
            final var cipher = Cipher.getInstance(NoiseConstants.CIPHER_ALGORITHM);
            cipher.init(
                mode,
                new SecretKeySpec(key, "ChaCha20"),
                new IvParameterSpec(nonceBytes())
            );
            cipher.updateAAD(associatedData);
            final var output = cipher.doFinal(input);
            nonce += 1;
            return output;
        } catch (GeneralSecurityException exception) {
            throw new EspHomeProtocolException("Noise cipher operation failed", exception);
        }
    }

    private byte[] nonceBytes() {
        final var bytes = new byte[12];
        var remaining = nonce;
        for (var index = 4; index < 12; index++) {
            bytes[index] = (byte) (remaining & 0xffL);
            remaining >>>= 8;
        }
        return bytes;
    }
}

File: services/lib-esphome-client/src/main/kotlin/io/github/arhor/esphome/client/internal/noise/NoiseCipherState.kt
package io.github.arhor.esphome.client.internal.noise

import io.github.arhor.esphome.client.EspHomeProtocolException
import java.security.GeneralSecurityException
import javax.crypto.Cipher
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec

internal class NoiseCipherState {

    private var key: ByteArray? = null
    private var nonce: Long = 0

    fun initializeKey(key: ByteArray?) {
        this.key = key?.copyOf()
        nonce = 0
    }

    fun hasKey(): Boolean =
        key != null

    fun encryptWithAd(associatedData: ByteArray, plaintext: ByteArray): ByteArray {
        val currentKey = key ?: return plaintext
        if (plaintext.size + NoiseConstants.AUTH_TAG_LENGTH > NoiseConstants.NOISE_MAX_MESSAGE_LENGTH) {
            throw EspHomeProtocolException("Noise plaintext is too large: ${plaintext.size} bytes")
        }

        return runCipher(Cipher.ENCRYPT_MODE, currentKey, associatedData, plaintext)
    }

    fun decryptWithAd(associatedData: ByteArray, ciphertext: ByteArray): ByteArray {
        val currentKey = key ?: return ciphertext
        if (ciphertext.size < NoiseConstants.AUTH_TAG_LENGTH) {
            throw EspHomeProtocolException("Noise ciphertext is too short")
        }

        return runCipher(Cipher.DECRYPT_MODE, currentKey, associatedData, ciphertext)
    }

    private fun runCipher(mode: Int, currentKey: ByteArray, associatedData: ByteArray, input: ByteArray): ByteArray {
        try {
            val cipher = Cipher.getInstance(NoiseConstants.CIPHER_ALGORITHM)
            cipher.init(
                mode,
                SecretKeySpec(currentKey, "ChaCha20"),
                IvParameterSpec(nonceBytes()),
            )
            cipher.updateAAD(associatedData)
            val output = cipher.doFinal(input)
            nonce += 1
            return output
        } catch (exception: GeneralSecurityException) {
            throw EspHomeProtocolException("Noise cipher operation failed", exception)
        }
    }

    private fun nonceBytes(): ByteArray {
        val bytes = ByteArray(12)
        var remaining = nonce
        for (index in 4 until 12) {
            bytes[index] = (remaining and 0xffL).toByte()
            remaining = remaining ushr 8
        }
        return bytes
    }
}
